### ISSUE TITLE
Unicode string is truncated when python uses 2 byte unicode. (migrated from Trac #2081)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#3628]
Original ticket http://projects.scipy.org/numpy/ticket/2081
Reported 2012-03-14 by trac user taldcroft, assigned to unknown.

Using numpy devel with Python 3.2 gives an unexpected result trying to initialize an array from numpy string objects.  This is a regression from previous behavior (devel prior to about two weeks ago and 1.6.1).

devel:

```
>>> d = np.array(['a'])
>>> np.asarray([d[0]])
array([''], 
      dtype='<U0')
```

In numpy 1.6.1:

```
>>> d = np.array(['a'])
>>> np.asarray([d[0]])
array(['a'], 
      dtype='<U1')
```

If one uses a longer string like "abcd" instead of "a" then the result is "ab", so it looks like the actual output length is // 2 from the desired length.  Perhaps this is related to:

https://github.com/numpy/numpy/commit/91f87e1f613630ff0ad9864017f059afcd6e57f1

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2012-03-15

I can't duplicate this here:

```
$charrisatmention:f16 ~$ python3
Python 3.2.1 (default, Jul 11 2011, 18:54:42) 
[GCC 4.6.1 20110627 (Red Hat 4.6.1-1)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> d = np.array(['a'])
>>> np.asarray(d[0])
array('a', 
      dtype='<U1')
>>> d = np.array([b'a'])
>>> np.asarray(d[0])
array(b'a', 
      dtype='|S1')
>>> np.__version__
'1.7.0.dev-cfee94e'
>>> d = np.array(['abcd'])
>>> np.asarray(d[0])
array('abcd', 
      dtype='<U4')
```

Can you try with latest numpy master and provide more info on the OS/Platform, etc., if it still fails.

**Comment by thouis** [comment]
Comment in Trac by trac user taldcroft, 2012-03-15

I'm using linux CentOS-5 x86-64.  In my example I did

```
np.asarray([d[0]])
```

where the attempt to duplicate uses:

```
np.asarray(d[0])  # missing enclosing [] brackets
```

Here is my Python and version info:

```
(astropy32-numpydev)ccosmos$ python
ActivePython 3.2.0.0 (ActiveState Software Inc.) based on
Python 3.2 (r32:88445, Feb 21 2011, 11:25:11) 
[GCC 4.0.2 20051125 (Red Hat 4.0.2-8)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> np.__version__
'1.7.0.dev-cfee94e'
```

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2012-03-16

I still can't reproduce this:

```
>>> import numpy as np
>>> d = np.array(['a'])
>>> np.asarray([d[0]])
array(['a'], 
      dtype='<U1')
>>> d = np.array(['abcd'])
>>> np.asarray([d[0]])
array(['abcd'], 
      dtype='<U4')
>>> np.__version__
'1.7.0.dev-0258f19'
```

I wonder if this is a bug in python 3.2.0? I'm running 3.2.1 and gcc 4.6.1. Just to be sure it's nothing, can you remove the build and install directories and do a clean install.

**Comment by thouis** [comment]
Comment in Trac by trac user taldcroft, 2012-03-16

Replying to [comment:3 charris]:

> I still can't reproduce this:
> 
> {{{
> 
> > > > import numpy as np
> > > > d = np.array(['a'])
> > > > np.asarray([d[0]])
> > > > array(['a'], 
> > > >       dtype='<U1')
> > > > d = np.array(['abcd'])
> > > > np.asarray([d[0]])
> > > > array(['abcd'], 
> > > >       dtype='<U4')
> > > > np.**version**
> > > > '1.7.0.dev-0258f19'
> 
> }}}
> 
> I wonder if this is a bug in python 3.2.0? I'm running 3.2.1 and gcc 4.6.1. Just to be sure it's nothing, can you remove the build and install directories and do a clean install.

I installed a fresh copy of Python 3.2.2.3 from ActiveState community edition, deleted the build/ directory in my git numpy repo, set my PATH to start with the ActivePython bin directory, and ran "python3 setup.py install".  I still see the same problem (see below).  

I should note that this all started with the same problem that showed up in a Shining Panda Jenkins instance using Python 3.2 and numpy-dev.  I think that is running Debian (but I am not certain of details), so this issue is probably not restricted to CentOS-5 or ActivePython.

```
ActivePython 3.2.2.3 (ActiveState Software Inc.) based on
Python 3.2.2 (default, Sep  8 2011, 12:20:18) 
[GCC 4.0.2 20051125 (Red Hat 4.0.2-8)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np
>>> np.__version__
'1.7.0.dev-cfee94e'
>>> d = np.array(['a'])
>>> np.asarray([d[0]])
array([''], 
      dtype='<U0')
```

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2012-03-16

Hmm, I suspect it's either the compiler or a ucs2/ucs4 problem, most likely the latter. Can you check the PY_UNICODE_TYPE?

```
$charrisatmention:f16 ~$ grep -in Py_UNICODE_SIZE /usr/include/python3.2mu/pyconfig-64.h
984:#define Py_UNICODE_SIZE 4
```

**Comment by thouis** [comment]
Comment in Trac by trac user taldcroft, 2012-03-16

Replying to [comment:5 charris]:

> Hmm, I suspect it's either the compiler or a ucs2/ucs4 problem, most likely the latter. Can you check the PY_UNICODE_TYPE?
> 
> {{{
> $charrisatmention:f16 ~$ grep -in Py_UNICODE_SIZE /usr/include/python3.2mu/pyconfig-64.h
> 984:#define Py_UNICODE_SIZE 4
> }}}

I do see something different, though note there is no pyconfig-64.h file in the distribution:

```
ccosmos$ grep -in Py_UNICODE_SIZE include/python3.2m/pyconfig.h 
978:#define Py_UNICODE_SIZE 2
```

(This is from the root of my Python 3.2.2.3 installation).

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2012-03-16

Thanks, that looks like the problem. Now to figure out the fix...

**Comment by thouis** [comment]
Attachment in Trac by atmention:charris, 2012-03-17: [unicode-patch.patch](http://projects.scipy.org/numpy/attachment/ticket/2081/unicode-patch.patch)

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2012-03-17

Could you try the attached patch to see if it fixes the problem?

**Comment by thouis** [comment]
Comment in Trac by trac user taldcroft, 2012-03-17

Replying to [comment:8 charris]:

> Could you try the attached patch to see if it fixes the problem?

That worked for me on Python 3.2.2.3 and I now get the expected result.  Thanks!

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2012-03-17

Fixed in 32a4a7d. Thanks for the report.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2012-04-30
